### PR TITLE
`Communication`: Fix clipping issues on mobile devices

### PR DESCRIPTION
**Pull request by PaRangger**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

The communication module is currently not optimized for mobile devices, resulting in UI elements being clipped. While native applications are available, some users may still attempt to access the module via smartphones. Addressing the clipping issue will enhance the module’s accessibility and improve the overall user experience.

### Description

I resolved the clipping issues by overlaying the sidebar, search, messages, and answer thread. Additionally, I introduced a button in the conversation title to navigate back to the sidebar. When performing a search, the sidebar now automatically hides. This hide-and-show functionality of the sidebar is managed through a newly implemented event service. To avoid confusion, the "All Messages" tab is hidden on smartphones.

Most of the changes were made in CSS. However, for the sidebar functionality, the layout service is checked to determine if the device is a smartphone. These changes are designed to have no impact on the desktop version of the application.

This update serves as a quick fix to ensure basic usability of the module, including channel selection, reading and writing messages, and viewing answer threads.

Known Bugs for followup PRs:
1. The emoji bar is clipping on some screens.
2. On iOS, the text-cursor shows a litte blue mark on top.

### Steps for Testing

Prerequisites:
- 1 Student/Instructor/Tutor
- 1 Course with communication enabled
- Access the page via smartphone, simulator or in the smartphone debug window of a browser.

Bildschirmfoto 2024-10-12 um 23 45 21

1. Log in to Artemis
3. Navigate to the course
4. In the sidebar, navigate to the communication tab.
5. Check if the module is navigatable and the main use cases of the module are possible.

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| course-conversations.component.ts | 89.2% | ✅ |
| course-wide-search.component.ts | 97.36% | ✅ |
| conversation-header.component.ts | 95.08% | ✅ |
| course-overview.component.ts | 86.89% | ✅ |
| course-sidebar.service.ts | 100% | ✅ |
| layout.service.ts | 93.33% | ✅ |

### Screenshots

Bildschirmfoto 2024-10-12 um 23 56 13

Bildschirmfoto 2024-10-12 um 23 56 00

Bildschirmfoto 2024-10-12 um 23 56 26

## Release Notes

- **New Features**
  - Introduced mobile detection to disable tooltips for mobile users.
  - Enhanced sidebar functionality with a new service for managing visibility and state.
  - Improved UI responsiveness with dynamic class bindings for conversation and sidebar states.
  - Added a message input section for creating new messages based on conversation context.
  - Integrated new styles and responsive design elements for better mobile experience.
  - Implemented a mobile disclaimer in the course-wide search interface.
  - Added new localization keys for mobile guidance in the German language.

- **Bug Fixes**
  - Adjusted sidebar toggle behavior to ensure correct state management on mobile devices.

- **Style**
  - Updated styles for better responsiveness, including new font size classes and removal of `!important` directive.
  - Enhanced styling capabilities with Bootstrap SCSS imports and new responsive design elements.

- **Tests**
  - Expanded test coverage for sidebar functionality and component interactions, ensuring robust behavior across scenarios.

### CODE REVIEW

**Review by sachmii — COMMENTED** [review]
Tested on TS5 on my smartphone and using the smartphone simulator in the browser.
I really like the transition when clicking on a communication channel and using the button to leave the channel and go back to the overview; it doesn't look abrupt.
However, when I add a channel to the _Favorites_ section, the previously opened channel opens again.
If I haven't opened any channel prior to adding one to my _Favorites_ section, the issue does not occur.

**Comment by PaRangger** [comment]
> Tested on TS5 on my smartphone and using the smartphone simulator in the browser. I really like the transition when clicking on a communication channel and using the button to leave the channel and go back to the overview; it doesn't look abrupt. However, when I add a channel to the _Favorites_ section, the previously opened channel opens again. If I haven't opened any channel prior to adding one to my _Favorites_ section, the issue does not occur.

Good catch with the favorites! Should be fixed with this [commit](https://github.com/ls1intum/Artemis/pull/9460/commits/bb1e097d0f16109d75ed7d16bcab0d4b84759531). 👍🏼

**Review by HawKhiem — COMMENTED** [review]
I tested on TS1 using the Iphone 14 pro max simulator in Chrome. Love the changes. No issues were found. One thing I did notice was, when I tried to hide a channel which is being marked as favorite, that channel stays in favorites instead of being moved to hidden. Although I think this has been discussed/resolved in another pull request. (PR #9473)

**Comment by zagemello** [comment]
I tested it on my laptop using the smartphone emulator and found no errors. I was also able to favourite without issue, and nothing irregular occurred when removing the favourite tag.

**Review by zagemello — COMMENTED** [review]
I tested it on my laptop using the smartphone emulator and found no errors. I was also able to favourite without issue, and nothing irregular occurred when removing the favourite tag.

**Review by asliayk — DISMISSED** [review]
tested locally, and it works as expected. I was able to navigate to the channels, threads, and sidebar view.

**Comment by ahbitaqu** [comment]
Tested on test-server2 using a pixel 7 and found a couple of minor issues:

1. Scrolling on the course navigation page only works consistently if you grab around the course tiles. Otherwise it's kinda random?!? (image 1)
2. When trying to react with an emoji or trying to tag someone / channel, the popup is not visible entirely (image 2)
3. The Sidebars are distorted (image 3 is prod image 4 is test server)

Apart from that, nice work. This feature was really overdue especially for tutors since the android app has too many issues to use properly as of right now.

Screenshots:
1. 
share_5193675039747908128

2.
Screenshot 2024-10-20 at 11-31-44 Communication Practical Course Interactive Learning WS24_25

3.
Screenshot_20241020-113731

4. 
Screenshot_20241020-113743

**Comment by PaRangger** [comment]
> Tested on test-server2 using a pixel 7 and found a couple of minor issues:
> 
> 1. Scrolling on the course navigation page only works consistently if you grab around the course tiles. Otherwise it's kinda random?!? (image 1)
> 2. When trying to react with an emoji or trying to tag someone / channel, the popup is not visible entirely (image 2)
> 3. The Sidebars are distorted (image 3 is prod image; 4 is test server)
> 
> Apart from that, nice work. This feature was really overdue especially for tutors since the android app has too many issues to use properly as of right now.
> 
> Screenshots: 1. share_5193675039747908128
> 
> 
> Screenshot 2024-10-20 at 11-31-44 Communication Practical Course Interactive Learning WS24_25
> 
> 
> Screenshot_20241020-113731
> 
> 
> Screenshot_20241020-113743

Hey thank you for testing! For your feedback:

1. I noticed this too, but since it wasn't connected to the communication module per se, i did not wanna touch it in this PR. However i think it'll be good to work on this in a followup.
2. Thank you for noticing. I tried to fix it on mobile screens with a small media query.
3. The sidebar is supposed to be over the entire screen now, and the smaller elements are from other pr's. I think this is intentional. Unless you mean something different?

**Review by SindiBuklaji — COMMENTED** [review]
I tested on TS3 using a Samsung Galaxy S22. 

- I see it was mentioned before but fyi I am also having difficulties scrolling in the courses page. You need to touch very specifically in the course, for it to work. 
- For the moment these two buttons have the same functionality. Is that the purpose? 
Screenshot_20241020_171402_Edge

- When I start typing, the first letter is always hidden. Only when I press space does the full word show up. 
image


- (not sure if this is relevant here but)  when I rotate the phone horizontally, I can't scroll at all here on this side: 
image

**Comment by PaRangger** [comment]
> I tested on TS3 using a Samsung Galaxy S22.
> 
> * I see it was mentioned before but fyi I am also having difficulties scrolling in the courses page. You need to touch very specifically in the course, for it to work.
> * For the moment these two buttons have the same functionality. Is that the purpose?
>   Screenshot_20241020_171402_Edge
> * When I start typing, the first letter is always hidden. Only when I press space does the full word show up.
>   image
> * (not sure if this is relevant here but)  when I rotate the phone horizontally, I can't scroll at all here on this side:
>   image

Thanks for testing and the feedback! 😄 Concerning your remarks:

1. The issue you described definitely needs to be fixed, however i thought it is not within the scope of this PR.
2. The 2 buttons that do the same thing are meant like that. I think one of the main issues is the navigation. While i wanted to keep the main components of the desktop version there, i felt like it was still hard to "go back" to the sidebar. Thats why I introduced the additional button.
3. The typing is actually a first time this comes up and should definitely not be like this. Which version of android and what browser did you use? Could you verify if its also like this on the production client? I cannot replicate it on my android phone unfortunately.
4. And lastly, i think locking the phone into portrait mode would make sense, but it could affect other components, so i did not wanna mess with meta tags.

**Comment by ahbitaqu** [comment]
>     2. Thank you for noticing. I tried to fix it on mobile screens with a small media query.

unfortunately this still doesn't work, now the popup appears on the other end of the screen 🙃 (image 1)

> 
>     3. The sidebar is supposed to be over the entire screen now, and the smaller elements are from other pr's. I think this is intentional. Unless you mean something different?

If you look at the picture on the top left you can see that the entire side bar feels like it has been squished vertically. Also the channel fields are much smaller on the y-Axis (image 2; image 3 is current prod)

Image 1:
Screenshot_20241020-220843

Image 2:
grafik

Image 3:
grafik

**Comment by PaRangger** [comment]
> unfortunately this still doesn't work, now the popup appears on the other end of the screen 🙃 (image 1)

Ok not good 😅 I reverted the change - its a bit hard to deal with the emoji bar since all of the positioning is done within a 3rd party package. I'll try to figure this out in a follow-up PR.

> If you look at the picture on the top left you can see that the entire side bar feels like it has been squished vertically (picture is oval shaped). Also the channel fields are much smaller on the y-Axis (image 2; image 3 is current prod)

Oh now I see you are right. The pixel count on the image for that specific course is not square 😅 I'm unsure how these are set up but it might be a testserver setup thing.

Bildschirmfoto 2024-10-21 um 11 36 48

The smaller sidebar items are part of a sidebar redesign in another PR, i think this was intentional and should be on prod now.

**Comment by SindiBuklaji** [comment]
> > I tested on TS3 using a Samsung Galaxy S22.
> > 
> > * I see it was mentioned before but fyi I am also having difficulties scrolling in the courses page. You need to touch very specifically in the course, for it to work.
> > * For the moment these two buttons have the same functionality. Is that the purpose?
> >   Screenshot_20241020_171402_Edge
> > * When I start typing, the first letter is always hidden. Only when I press space does the full word show up.
> >   image
> > * (not sure if this is relevant here but)  when I rotate the phone horizontally, I can't scroll at all here on this side:
> >   image
> 
> Thanks for testing and the feedback! 😄 Concerning your remarks:
> 
> 1. The issue you described definitely needs to be fixed, however i thought it is not within the scope of this PR.
> 2. The 2 buttons that do the same thing are meant like that. I think one of the main issues is the navigation. While i wanted to keep the main components of the desktop version there, i felt like it was still hard to "go back" to the sidebar. Thats why I introduced the additional button.
> 3. The typing is actually a first time this comes up and should definitely not be like this. Which version of android and what browser did you use? Could you verify if its also like this on the production client? I cannot replicate it on my android phone unfortunately.
> 4. And lastly, i think locking the phone into portrait mode would make sense, but it could affect other components, so i did not wanna mess with meta tags.

I am testing on Microsoft Edge browser and my android version is 14. I also tested on the published version of Artemis and I can see the error there as well.

**Review by janthoXO — CHANGES_REQUESTED** [review]
The waiting cursor looks kind of odd. There a blue dot at the top, which i think is the normal waiting cursor. So it is miss placed to the real one.
https://github.com/user-attachments/assets/b21c7a06-eebe-462a-a686-8f0bc0cd5d05

Normally it should look like this:

https://github.com/user-attachments/assets/bb98447f-71e6-48e6-a71a-4fa8ea63e210

**Review by janthoXO — CHANGES_REQUESTED** [review]
Theres a bug where you can not scroll in a channel anymore. As far as i can tell it happens when tapping outside the channel container or scroll outside the container and then try to scroll in the channel again. When pressing in the container again, the scrolling works again. 
Still this can confuse users and should be adjusted.
https://github.com/user-attachments/assets/8e68e913-f355-43d6-b4ee-831d5af1aaa9

**Review by Malekos74 — APPROVED** [review]
### Server tested on: Server 3
Works as expected, no overflows nowhere.
**Minor notice:**
I think the problem cited above about the icon is not caused by this PR. The first couse has a logo that is slightly rectangular, whereas the second has a perfectly square logo.

image


image

**Comment by PaRangger** [comment]
> Theres a bug where you can not scroll in a channel anymore. As far as i can tell it happens when tapping outside the channel container or scroll outside the container and then try to scroll in the channel again. When pressing in the container again, the scrolling works again. Still this can confuse users and should be adjusted. https://github.com/user-attachments/assets/8e68e913-f355-43d6-b4ee-831d5af1aaa9

Thank you for the feedback! I tried to fix the overflow behaviour in this [commit](https://github.com/ls1intum/Artemis/pull/9460/commits/a693dbc7556d474b7574d07dc9a3d225db16839c). I hope now it works as expected 👍🏼

> The waiting cursor looks kind of odd. There a blue dot at the top, which i think is the normal waiting cursor. So it is miss placed to the real one. https://github.com/user-attachments/assets/b21c7a06-eebe-462a-a686-8f0bc0cd5d05

I have been looking for a fix to this issue but it seems quite involved. I think i can make a standalone issue to solve this, since this PR is supposed to be a "quick" fix to make the module generally usable.

> I cannot find a way to let the exercise/lecture/communication sidebar extend. This means the user has to rely on the icons, which can be not accessible for first time users. As they dont know the icons and functionalities of artemis yet.

The exercise/lecture sidebar should not be part of this PR. For the communication I tried to add a chevron next to the channel name as well as the search, just like in slack. I was hoping this, in addition to the usual sidebar button will be enough for the user to know how to navigate.

File: src/main/webapp/app/overview/course-conversations/course-conversations.component.ts [hunk 1]
```diff
@@ -138,6 +146,8 @@ export class CourseConversationsComponent implements OnInit, OnDestroy {
         private courseOverviewService: CourseOverviewService,
         private modalService: NgbModal,
         private profileService: ProfileService,
+        private courseSidebarService: CourseSidebarService,
```
Comment (cremertim) [inline src/main/webapp/app/overview/course-conversations/course-conversations.component.ts:149]: can we use inject for the new services here?

File: src/main/webapp/app/overview/course-conversations/layout/conversation-header/conversation-header.component.ts [hunk 2]
```diff
@@ -42,13 +43,15 @@ export class ConversationHeaderComponent implements OnInit, OnDestroy {
...
     faSearch = faSearch;
+    faChevronLeft = faChevronLeft;
 
     constructor(
         private modalService: NgbModal,
         // instantiated at course-conversation.component.ts
         public metisConversationService: MetisConversationService,
         public conversationService: ConversationService,
         private metisService: MetisService,
+        private courseSidebarService: CourseSidebarService,
```
Comment (cremertim) [inline src/main/webapp/app/overview/course-conversations/layout/conversation-header/conversation-header.component.ts:54]:
```suggestion        
  private courseSidebarService= inject(CourseSidebarService);
```

File: src/main/webapp/app/overview/course-overview.component.ts [hunk 3]
```diff
@@ -202,10 +206,22 @@ export class CourseOverviewComponent implements OnInit, OnDestroy, AfterViewInit
         private profileService: ProfileService,
         private modalService: NgbModal,
         private examParticipationService: ExamParticipationService,
+        private courseSidebarService: CourseSidebarService,
```
Comment (cremertim) [inline src/main/webapp/app/overview/course-overview.component.ts:209]:
```suggestion
        courseSidebarService = inject(CourseSidebarService)
```

File: src/main/webapp/i18n/de/metis.json [hunk 4]
```diff
@@ -92,7 +92,9 @@
                 "RANDOM": "Sonstiges",
                 "ANNOUNCEMENT": "Ankündigung",
                 "allPublicMessages": "Alle öffentlichen Nachrichten",
-                "searchResults": "Suchergebnisse für {{ search }}"
+                "searchResults": "Suchergebnisse für {{ search }}",
+                "mobileDisclaimer": "Bitte wähle erst eine Konversation aus.",
```
Comment (cremertim) [inline src/main/webapp/i18n/de/metis.json:96]:
```suggestion
                "mobileDisclaimer": "Bitte erst eine Konversation auswählen.",
```
I think this easier to read

File: src/main/webapp/app/overview/course-conversations/course-wide-search/course-wide-search.component.ts [hunk 5]
```diff
@@ -74,6 +76,7 @@ export class CourseWideSearchComponent implements OnInit, AfterViewInit, OnDestr
         public metisConversationService: MetisConversationService, // instance from course-conversations.component
         private formBuilder: FormBuilder,
         public cdr: ChangeDetectorRef,
+        private courseSidebarService: CourseSidebarService,
```
Comment (cremertim) [inline src/main/webapp/app/overview/course-conversations/course-wide-search/course-wide-search.component.ts:79]:
```suggestion
        private courseSidebarService = inject(CourseSidebarService)
```

**Review by cremertim — CHANGES_REQUESTED** [review]
Code changes look good to me, some minor adjustments to fullfill the client guidelines

File: src/main/webapp/app/overview/course-conversations/course-wide-search/course-wide-search.component.ts (hunk 5 above)
Comment (PaRangger) [reply src/main/webapp/app/overview/course-conversations/course-wide-search/course-wide-search.component.ts]: Thanks for the feedback! I addressed it in this [commit](https://github.com/ls1intum/Artemis/pull/9460/commits/068c060c09808233eb33695f12486d82466e22dd).

File: src/main/webapp/i18n/de/metis.json (hunk 4 above)
Comment (PaRangger) [reply src/main/webapp/i18n/de/metis.json]: I agree, I addressed it in this [commit](https://github.com/ls1intum/Artemis/pull/9460/commits/068c060c09808233eb33695f12486d82466e22dd). 👍🏼

**Review by FelberMartin — APPROVED** [review]
Tested the functionality as described in the PR, works as expected. The main chatting functionality is now accessible via the mobile browser.

This is an important step to improve usability, good work 👌

**Comment by Malekos74** [comment]
I couldn't deploy on server 1 nor 4 even though they were free, I will try again later.
o7

**Review by sachmii — APPROVED** [review]
Tested on TS5, works fine.,

**Review by Malekos74 — DISMISSED** [review]
Tested on test Server 5
Looks great 🫡

**Review by asliayk — APPROVED** [review]
tested locally and everything works fine

**Pull request merged**